src/main/java/build/buildfarm/worker/shard/ShardWorkerContext.java

```diff
@@ -90,18 +96,22 @@
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.Comparator;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Random;
 import java.util.Set;
 import java.util.Stack;
 import java.util.UUID;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
 class ShardWorkerContext implements WorkerContext {
   private static final Logger logger = Logger.getLogger(ShardWorkerContext.class.getName());
 
+  private final com.google.common.cache.LoadingCache<String, Instance> workerStubs;
```

Comment:
why the fqcn here? See the below comment about how I don't think this class should be made aware of workerStubs at all.